Makefile

```diff
@@ -54,6 +54,7 @@ GO_APIDIFF := $(TOOLS_DIR)/$(GO_APIDIFF_BIN)
 # Binaries.
 # Need to use abspath so we can invoke these from subdirectories
 KUSTOMIZE := $(abspath $(TOOLS_BIN_DIR)/kustomize)
+COUNTERFEITER := $(abspath $(TOOLS_BIN_DIR)/counterfeiter)
```

Comment:
I'd rather avoid using mocks to test controller behavior. Instead, we can test the full controller behavior adding an integration test in the current MachineSet ginkgo tests

Reply:
The idea here is that the remediation is not the controller's behavior, it's the responsibility of the remediator that the controller depends on but does not need to understand. It's a contract that you call `Remediate` and give it machines and it will tell you whether any remediation happened or not.

A unit test is for testing one unit at a time, so here we can assert both sides of the contract but not the interdependency. I agree we should have some tests that do test these things together, I just am arguing they're not unit tests. What I am trying to do here is break up code that is in my opinion way too coupled. The controllers are the entrypoint to the system but that does not mean that we can't have units of code that are not just subcomponents of the controller.

Reply:
> It's a contract that you call Remediate and give it machines and it will tell you whether any remediation happened or not.

The contract, from an API point of view, states that the only expectation for the owning-controller is to delete the Machine. Code shouldn't be part of any contract, one example is controllers not written in Go, or not using Cluster API as a library. They should be able to function by looking at the data and make decisions, ultimately satisfy the API contract expectations.

While it's fine to have unit tests in the exposed remediation functions, I'd like to avoid to test mocked tests when we have the ability to run an integration test which is going to be much closer to real world usage.

How we structure code is a different thing, the remediation code can be tested separately.

Reply:
I am strongly opposed to mocks in the majority of cases. Speaking from experience, mocking behavior makes your tests extremely brittle. If you change the internals of your implementation, this frequently requires you to adjust the mocks' expectations. This may seem reasonable and easy if it's a couple of days after the initial implementation is committed to the repo, if you (or worse, someone else) come back weeks later, it's often very difficult to page things back in to memory. This is made worse if you have complicated test logic (if xyz set, expect mock to do this; else don't expect it) and tons of tests.

Reply:
This is the result of poorly written code and tests and not the fault of the mocks. I'm not interested in fighting this fight anymore so I'll remove them.